cpp/src/arrow/compute/registry.cc

```diff
@@ -115,6 +116,19 @@ static std::unique_ptr<FunctionRegistry> CreateBuiltInRegistry() {
   RegisterVectorNested(registry.get());
   RegisterVectorSort(registry.get());
 
+  // SIMD functions
+  auto cpu_info = arrow::internal::CpuInfo::GetInstance();
+#if defined(ARROW_HAVE_RUNTIME_AVX2)
+  if (cpu_info->IsSupported(arrow::internal::CpuInfo::AVX2)) {
+    RegisterScalarAggregateSumAvx2(registry.get());
```

Comment:
Does this mean the sum function gets defined twice for certain input types? I'm not sure the behaviour of overriding the previous definitions is well-defined. @wesm What do you think?

Reply:
@jianxind is using the `allow_overwrite` option in `AddFunction`. 

It's a little bit coarse to do things this way -- I do not see this as the preferred long-term solution. Instead, we should add all the kernel variants to the same `Function` instance and then the `Dispatch*` methods should select the kernel with the maximum SIMD level as set on the Kernel object. That was the idea of the `simd_level` parameter

https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernel.h#L562

If you don't do this, then if you want to apply SIMD optimizations in even one variant of a kernel then it's a bit of a pain

Reply:
I don't note the simd_level parameter before. We can put it a further optimization then, I guess there's still some jobs to make simd_level work as expected.

Reply:
https://issues.apache.org/jira/browse/ARROW-9398 created to track the SIMD dispatch under function instance